Remotion/Data/DomainObjects/Queries/ConfigurationLoader/QueryDefinitionLoader.cs

```diff
@@ -0,0 +1,72 @@
+﻿// This file is part of the re-motion Core Framework (www.re-motion.org)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+//
+// The re-motion Core Framework is free software; you can redistribute it
+// and/or modify it under the terms of the GNU Lesser General Public License
+// as published by the Free Software Foundation; either version 2.1 of the
+// License, or (at your option) any later version.
+//
+// re-motion is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+//
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using Remotion.Data.DomainObjects.Configuration;
+using Remotion.Data.DomainObjects.Persistence;
+using Remotion.Data.DomainObjects.Queries.Configuration;
+using Remotion.ServiceLocation;
+using Remotion.Utilities;
+
+namespace Remotion.Data.DomainObjects.Queries.ConfigurationLoader
+{
+  /// <summary>
+  /// Default implementation of the <see cref="IQueryDefinitionLoader"/> interface.
+  /// </summary>
+  [ImplementationFor(typeof(IQueryDefinitionLoader), Lifetime = LifetimeKind.Singleton)]
+  public class QueryDefinitionLoader : IQueryDefinitionLoader
+  {
+    private readonly IQueryFileFinder _queryFileFinder;
+
+    private readonly DoubleCheckedLockingContainer<ReadOnlyCollection<QueryDefinition>> _queries;
+
+    public QueryDefinitionLoader (IQueryFileFinder queryFileFinder)
+    {
+      ArgumentUtility.CheckNotNull(nameof(queryFileFinder), queryFileFinder);
+
+      _queryFileFinder = queryFileFinder;
+
+      _queries = new DoubleCheckedLockingContainer<ReadOnlyCollection<QueryDefinition>>(LoadAllQueryDefinitionsImplementation);
+    }
+
+    /// <inheritdoc />
+    public ReadOnlyCollection<QueryDefinition> LoadAllQueryDefinitions () => _queries.Value!;
```

Comment:
If we want to use an `IReadOnlyXXX` here instead of `ReadOnlyCollection<T>` I would suggest using `IReadOnlyList<T>` instead of the collection one. I feel like the collection variant does not add much over `IEnumerable<T>` which could then be used instead.

I personally, don't think it is necessary to wrap a created list with a `ReadOnlyCollection<QueryDefinition>` if we return an `IReadOnlyList<T>` but I know mike likes it so I would keep the creation and only change the return type.

Reply:
I have nothing against `IReadOnlyList<T>`, but I think that `IReadOnlyCollection<T>` adds a significant advantage in that it is already enumerated and can thus be checked by `ArgumentUtility.CheckNotNullOrItemsNull()`. In my opinion, `IEnumerable<T>` should only be used when lazy enumeration is a required feature.

With regards to wrapping, I had that discussion. If we don't keep the list, we can just return the `List<T>` because even if the caller casts from `IReadOnlyCollection<T>` and modifies it, it does not corrupt our internal state. In this case however, we keep the list and return it in all future calls, so we want to guard against that sort of caller misbehavior.